Source/Serilog.Exceptions/Core/ExceptionPropertiesBag.cs

```diff
@@ -0,0 +1,49 @@
+namespace Serilog.Exceptions.Core
+{
+    using System;
+    using System.Collections.Generic;
+    using Serilog.Exceptions.Filters;
+
+    internal class ExceptionPropertiesBag : IExceptionPropertiesBag
+    {
+        private readonly Type exceptionType;
+        private readonly IExceptionPropertyFilter filter;
+        private readonly Dictionary<string, object> properties = new Dictionary<string, object>();
+        private bool resultsCollected = false;
+
+        public ExceptionPropertiesBag(Type exceptionType, IExceptionPropertyFilter filter = null)
+        {
+            this.exceptionType = exceptionType;
+            this.filter = filter;
+        }
+
+        public IReadOnlyDictionary<string, object> GetResultDictionary()
+        {
+            this.resultsCollected = true;
+            return this.properties;
+        }
+
+        public void AddProperty(string key, object value)
+        {
+            if (key == null)
+            {
+                throw new ArgumentNullException(nameof(key), "Cannot add exception property without a key");
+            }
+
+            if (this.resultsCollected)
```

Comment:
Is there a reason for us to stop adding/removing properties at runtime?

Reply:
It is just an application of the fail-fast principle. If somebody is modifying result after it was collected and used in the resulting structure, there is something seriously wrong.

Reply:
I added a comment to document my thinking behind this invariant checking.